analyzers/src/SonarAnalyzer.CSharp/Rules/UseCachedRegex.cs

```diff
@@ -0,0 +1,114 @@
+                    && IsCorrectObjectType(objectCreationExpression, c.SemanticModel)
+                    && IsArgumentConstantOrReadOnly(objectCreationExpression, c.SemanticModel)
+                    && !IsCompliantAssignment(objectCreationExpression, c.SemanticModel))
+                {
+                    c.ReportIssue(Diagnostic.Create(Rule, objectCreationExpression.GetLocation()));
+                }
+            },
+            SyntaxKind.ObjectCreationExpression);
+
+    private static bool IsCompliantAssignment(ObjectCreationExpressionSyntax objectCreationExpression, SemanticModel model) =>
+        objectCreationExpression.Parent switch
+        {
+            AssignmentExpressionSyntax { RawKind: (int)SyntaxKind.SimpleAssignmentExpression } assignmentExpression =>
+                model.GetSymbolInfo(assignmentExpression.Left).Symbol is IFieldSymbol { IsReadOnly: true } or IPropertySymbol { IsReadOnly: true }
+                || IsAssignmentWithinIfStatement(assignmentExpression, model),
+            AssignmentExpressionSyntax { RawKind: (int)SyntaxKindEx.CoalesceAssignmentExpression } assignmentExpression =>
+                model.GetSymbolInfo(assignmentExpression.Left).Symbol is IFieldSymbol or IPropertySymbol,
+            BinaryExpressionSyntax
+            {
+                RawKind: (int)SyntaxKind.CoalesceExpression,
+                Parent: AssignmentExpressionSyntax { RawKind: (int)SyntaxKind.SimpleAssignmentExpression or (int)SyntaxKindEx.CoalesceAssignmentExpression } assignmentExpression
+            } coalesceExpression => coalesceExpression.Left.GetName() == assignmentExpression.Left.GetName()
+                                    && model.GetSymbolInfo(assignmentExpression.Left).Symbol is IFieldSymbol or IPropertySymbol,
+            _ => false
+        };
+
+    private static bool IsAssignmentWithinIfStatement(AssignmentExpressionSyntax assignmentExpression, SemanticModel model)
+    {
+        SyntaxNode node = assignmentExpression;
+        while (!node.Parent.IsKind(SyntaxKind.CompilationUnit))
+        {
+            if (node.Parent.IsKind(SyntaxKind.IfStatement))
+            {
+                break;
+            }
+
+            node = node.Parent;
+        }
+
+        return node.Parent is IfStatementSyntax { Condition: BinaryExpressionSyntax { RawKind: (int)SyntaxKind.EqualsExpression or (int)SyntaxKind.IsExpression } condition }
+               && ((assignmentExpression.Left.GetName() == condition.Left.GetName()
+                    && condition.Right is LiteralExpressionSyntax { RawKind: (int)SyntaxKind.NullLiteralExpression })
+                   || (assignmentExpression.Left.GetName() == condition.Right.GetName()
+                       && condition.Left is LiteralExpressionSyntax { RawKind: (int)SyntaxKind.NullLiteralExpression }))
+               && model.GetSymbolInfo(assignmentExpression.Left).Symbol is IFieldSymbol or IPropertySymbol;
```

Comment:
I think this is a bit complicated, from a cognitive point of view. 
You could extract the logic of left vs right in a local function and call it twice, something like:

```csharp
var check = ChecksForNullOnAssigned(condition.Left, condition.Right)
                       || ChecksForNullOnAssigned(condition.Right, condition.Left);


            bool ChecksForNullOnAssigned(ExpressionSyntax first, ExpressionSyntax second) =>
                assignmentExpression.Left.GetName() == first.GetName()
                && second is LiteralExpressionSyntax literal
                && literal.IsKind(SyntaxKind.NullLiteralExpression);
```

Also, to make it less nested, you can turn the first condition into two:
```csharp
node.Parent is IfStatementSyntax { Condition: BinaryExpressionSyntax condition }
                   && condition.IsAnyKind(SyntaxKind.EqualsExpression, SyntaxKind.IsExpression)
```

In total:

```csharp
return node.Parent is IfStatementSyntax { Condition: BinaryExpressionSyntax condition }
       && condition.IsAnyKind(SyntaxKind.EqualsExpression, SyntaxKind.IsExpression)
       && (ChecksForNullOnAssigned(condition.Left, condition.Right)
           || ChecksForNullOnAssigned(condition.Right, condition.Left))
       && model.GetSymbolInfo(assignmentExpression.Left).Symbol is IFieldSymbol or IPropertySymbol;

bool ChecksForNullOnAssigned(ExpressionSyntax first, ExpressionSyntax second) =>
    assignmentExpression.Left.GetName() == first.GetName()
    && second is LiteralExpressionSyntax literal
    && literal.IsKind(SyntaxKind.NullLiteralExpression);
```